ra/registration-authority.go

```diff
@@ -339,6 +347,36 @@ func (ra *RegistrationAuthorityImpl) NewCertificate(req core.CertificateRequest,
 		return emptyCert, err
 	}
 
+	// Check total certificate rate limiting
+	ra.tiMu.RLock()
+	now := ra.clk.Now()
+	if ra.totalIssuedCache == nil || ra.lastIssuedCount.Add(30*time.Minute).Before(now) {
+		// If we don't want reads to block each other we have to life with this horrible
+		// locking grossness (we could possibly use a goroutine that updates the cache
+		// and handles the write locking invisibly to this method but that seems kind
+		// of overkill to be honest)
+		ra.tiMu.RUnlock()
+		ra.tiMu.Lock()
```

Comment:
Also, instead of having to manage the RUnlock/RLock stuff in a bunch of a different places, it might be best to take the code that needs to call Lock/Unlock and put that in a separate function and make sure the lock is in the right state before and after.

Its currently very difficult to confirm which state the locks are at in any part of this code.